### ISSUE TITLE
StyleCop Violation SA12** not displayed in Visual Studio CodeFiles

### ISSUE THREAD

**Issue by agentschmitt** [DotNetAnalyzers/StyleCopAnalyzers#1939]
I have found some strange problem with some violations not displayed inside of visual studio code files.

The StyleCop Violation SA1200 and SA1210 are listed as warnings after building the project (in visual studio or with msbuild), but not displayed in the CodeFile in Visual Studio.
Also when i open the CodeFile, where the issue is listed, the issue dissapears from warning list, until i build the project again.

I am using StyleCopAnalyzers RC2, Visual Studio 2015 Update 1 (and Resharper 10.0.1).
I have also tryed to suspend Resharper to ensure it doesn't conflict with Stylecop, but thes didn't help.

**Comment by sharwell** [comment]
Can you post a file which reproduces this?

**Comment by nelsonghezzi** [comment]
Hi, I've seen SA1200 disappear since I've installed Update 1. No ruleset changes, just installed the update and the rule stopped from being reported, neither in the error list nor as a light bulb.

The versions I've used since the VS upgrade goes from beta017 to RC2, all with the same behavior.

Other SA12** rules (like 1202, 1203, 1205, 1210 and others) are still working in my project.

I was wondering if someone else has experimented this too, and this issue looks like the right place to report it.

BTW, kudos for all the effort put into this package!

**Comment by sharwell** [comment]
I installed Update 1 and I'm seeing the same behavior for SA1200.

It seems very inconsistent... will have to experiment further.

**Comment by sharwell** [comment]
@mavasani This issue is really strange, and I'm not quite sure how to handle it. The primary violation which we're seeing this is SA1200. The warnings are reported during an actual build, and also are passing in the unit tests under both Roslyn 1.0 and Roslyn 1.1. The strange behavior seems limited to the IntelliSense analysis and starts with Visual Studio 2015 Update 1.

**Comment by mavasani** [comment]
@sharwell Investigating now..

**Comment by mavasani** [comment]
@sharwell @agentschmitt I am unable to repro this - I tried it on a default C# console app on VS2015 Update1, with StyleCop.Analyzers 1.0.0-rc2 NuGet package and I am seeing 5 SA1200 warnings in the error list from build as well as intellisense.

Do you have a repro project/steps I can follow?

Additionally, is your build actually restoring the StyleCop analyzers NuGet package? If so, we fixed a related issue after Update1 (PR https://github.com/dotnet/roslyn/pull/7018)

**Comment by sharwell** [comment]
@mavasani

Steps to reproduce:

1. Check out sharwell/StyleCopAnalyzers@2ba137e737aa3cc5fdb8d6fc31a6194472c4b9a7
1. Open StyleCopAnalyzers.sln
1. Set the Error List filter to Build + IntelliSense
1. Restore NuGet packages
1. Restart Visual Studio and open StyleCopAnalyzers.sln again
1. Build (observe 51 SA1200 warnings)
1. Open each file containing a SA1200 warning reported by the build

Expected results:

Warnings are highlighted in the IDE, and a code fix can be executed on the line where the warning is reported.

Actual results:

The warnings do not appear in the editor at all. Despite being reported when the Error List is set to Build Only, they are not shown at all when the error list is set to Build + IntelliSense.

**Comment by mavasani** [comment]
@sharwell I was able to repro the issue with your solution - I have filed https://github.com/dotnet/roslyn/issues/7427 to track it.
Seems like analyzer is not getting called back for SyntaxKind.CompilationUnit for your solution. I am investigating further. Meanwhile I verified that changing the analyzer to register a SyntaxTree action works fine, and rule fires on StyleCop solution after upgrading it to locally built nupkg. You may want to either put this workaround or wait for the Roslyn bug to get fixed.

**Comment by sharwell** [comment]
@mavasani I'm seeing similar behavior if I edit **stylecop.json** to add the following `orderingRules` section:

```javascript
{
  "$schema": "https://raw.githubusercontent.com/DotNetAnalyzers/StyleCopAnalyzers/master/StyleCop.Analyzers/StyleCop.Analyzers/Settings/stylecop.schema.json",
  "settings": {
    "documentationRules": {
      // leave all these properties unchanged...
    },
    "orderingRules": {
      "usingDirectivesPlacement": "outsideNamespace"
    }
  }
}
```

I restarted Visual Studio after editing the file just to be sure.

In this case I expect the IDE to report SA1200 warnings in nearly every file in the solution. However, I am only seeing 18 warnings reported in 3 files. These warnings are reported in SA1200 by an action registered on SyntaxKind.NamespaceDeclaration, so it *seems* like it shouldn't be impacted by the CompilationUnit bug you already found.

**Comment by mavasani** [comment]
@sharwell @agentschmitt I found a simpler repro and this seems to be due to a race condition in Roslyn compiler layer: https://github.com/dotnet/roslyn/issues/7446.

**Comment by sharwell** [comment]
Closing as external; no workaround is available on this end.

**Comment by ScottGuymer** [comment]
Is there any know workaround to this issue? These analyzers become a little bit of a pain with this bug.

**Comment by sharwell** [comment]
@ScottGuymer Our workaround is [SA0000](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA0000Roslyn7446Workaround.md). For StyleCop Analyzers 1.1+ it will be automatically enabled (#2033), but for 1.0 you'll need to manually enable it for your project (see commit a48d687 as part of #2019).

**Comment by ScottGuymer** [comment]
Great thanks @sharwell it works a treat

**Comment by felickz** [comment]
FYI I am seeing this behavior for SA1200, SA1518, and SA1633  .. not sure what they have in common

**Comment by sharwell** [comment]
@felickz It's a bug in Roslyn itself that affects some analyzers and not others. See the workaround in an earlier comment for a resolution.

**Comment by mathmul** [comment]
> [@ScottGuymer](https://github.com/ScottGuymer) Our workaround is [SA0000](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA0000Roslyn7446Workaround.md).

404 - page not found
